### PR TITLE
Annotations on types

### PR DESCRIPTION
**Pull request by Cafeinoman**
Add the ability to retrieve annotations on each valid TypeInfo. So type annotations can be easily retrieved on template, even if we only have TypeInfo.

### CODE REVIEW

**Comment by vietj** [comment]
does this PR deprecates https://github.com/vert-x3/vertx-codegen/pull/127 ?

**Comment by Cafeinoman** [comment]
Yes, sorry, I should'nt do two PR.

**Comment by Cafeinoman** [comment]
Missing javadoc added. For test, there's a test on Class model too.

**Comment by vietj** [comment]
@Cafeinoman I'm curious, can you tell how this is used in your projects ? is it open sourced ?

**Comment by Cafeinoman** [comment]
Take a look at https://github.com/Cafeinoman/vertx-dao-generator . It's a work in progress, but the goal is to generate default DAO from annotated model at compile time, for Mongo, JDBC and Redis. I think I'll work, after that, on other project that can generate a lot of boilerplate code (REST services, default verticles, and so on).

**Comment by Cafeinoman** [comment]
any news?

**Comment by vietj** [comment]
I will continue the review soon

**Comment by vietj** [comment]
we should support annotation on methods for `DataObjectModel` as well

**Comment by Cafeinoman** [comment]
For all kind of methods? Property methods are already supported. I thought that data objects should only have data...

**Comment by vietj** [comment]
only for property accessors, I don't see any tests for that so

**Comment by Cafeinoman** [comment]
Property accessors are already supported, from PropertyInfo. Tests are present in DataObjectTest, from the AnnotatedDataObject class. Or there's something I misunderstood.

**Comment by vietj** [comment]
those are not new tests so ?

**Comment by Cafeinoman** [comment]
I wrote this test class when I started to work on this branch, and add some methods on DataObjectTest to keep all test in the same class.

**Comment by vietj** [comment]
actually looking at the PR, I do have a concern about this case:

```
@DataObject
@TheAnnotation(Foo.class)
public class Bar {
  public Bar(JsonObject json) {
  }
}

@DataObject
@TheAnnotation(Bar.class)
public class Foo {
  public Foo(JsonObject json) {
  }
}

public @interface TheAnnotation {
  Class<?> value();
}
```

that might run into a stack overflow error

**Comment by vietj** [comment]
this would be due to the fact that ClassTypeInfo can hold annotations that can have ClassTypeInfo as values and we get into these kind of problems.

if you look at the `TypeInfo` model, it does not expose purposely such traversal to avoid running into these kind of issues.

**Comment by Cafeinoman** [comment]
You're right, even if this case of annotation loop is a little bit strange. I'm looking for an alternative...

**Comment by Cafeinoman** [comment]
problem solved. Not very elegant, but it works.

**Comment by vietj** [comment]
@Cafeinoman we haven't finished to merged this PR, is that still useful to you ?

File: src/main/asciidoc/java/index.adoc [hunk 1]
```diff
@@ -690,7 +690,7 @@ The codegen processor _validates_ annotated Java program elements (i.e type decl
 
 Models are processed by https://en.wikisource.org/wiki/MVEL_Language_Guide[MVEL] templates, when a template is executed it gets access to implicit properties
 (i.e properties that are declared by the model).
-For ClassModel and DataObjectModel, annotations on class and methods can be accessed too. For ModuleModel, annotations on package can be accessed.
+For ClassModel and DataObjectModel, annotations on methods/properties can be accessed. For ModuleModel, annotations on package can be accessed. Annotation at type level can be retrieved for ClassModel, DataObjectModel, EnumModel and ProxyModel.
```
Comment (vietj) [inline src/main/asciidoc/java/index.adoc]: use backquotes for ClassModel, DataObjectModel, etc...

**Comment by vietj** [comment]
@Cafeinoman I'm actually seing an interest for this in the reactive postgres client to generate database tuples from data object for row mapping

**Comment by Cafeinoman** [comment]
Sorry for the delay. Yeah, it still useful for me. Not for the same thing as in the begining, but yeah. Anything to do on my side? Really lack of time...

**Comment by danielrivard** [comment]
Any plans to merge this pull request soon? I could use this functionality.

**Comment by vietj** [comment]
yes it could be but we need to resolve conflicts and properly review it

**Pull request merged**